returnn/torch/distributed.py

```diff
@@ -0,0 +1,135 @@
+import os
+import socket
+
+from contextlib import contextmanager
+import torch
+from torch.distributed.algorithms.join import Join
+
+from returnn.config import Config
+import returnn.frontend as rf
+
+
+class DistributedContext:
+    """
+    This class setups some helper functions for torch distributed training
+    """
+
+    def __init__(self, config):
+        """
+        :param Config config:
+        """
+        import torch.distributed as dist
+
+        dist.init_process_group("nccl")
+
+        self._config = config
+        self._local_rank = os.environ["LOCAL_RANK"]
+        self._local_size = os.environ["LOCAL_WORLD_SIZE"]
+        self._rank = dist.get_rank()
+        self._size = dist.get_world_size()
+
+        print(
+            "Torch distributed initialized. Hostname %s, pid %i, rank %i / size %i, local rank %s / local size %s."
+            % (socket.gethostname(), os.getpid(), self._rank, self._size, self._local_rank, self._local_size)
+        )
+
+    def local_rank(self):
+        """
+        :rtype: int
+        """
+        return self._local_rank
+
+    def rank(self):
+        """
+        :rtype: int
+        """
+        return self._rank
+
+    def size(self):
+        """
+        :rtype: int
+        """
+        return self._size
+
+
+_is_set_up = False
+_ctx = None  # type: typing.Optional[HorovodContext]
+
+
+def get_ctx(config=None):
+    """
+    :param Config|None config:
+    :returns: the global context if Torch distributed is enabled, or None otherwise.
+      If we did not setup the context yet, it will automatically create it.
+    :rtype: DistributedContext|None
+    """
+    global _is_set_up, _ctx
+    if _is_set_up:
+        return _ctx
+    if not config:
+        from returnn.config import get_global_config
+
+        config = get_global_config(raise_exception=False)
+        if not config:
+            return None
+    _is_set_up = True
+    if config.typed_value("torch_distributed") is None:
+        return None
+    _ctx = DistributedContext(config=config)
+    return _ctx
+
+
+@contextmanager
+def _ddp_ctx(pt_model):
```

Comment:
As you use this outside of this module, make it public. Also choose a more describing name, like `ddp_train_ctx` or so.